storey/sources.py

```diff
@@ -622,49 +621,78 @@ def __init__(
         WithUUID.__init__(self)
         if isinstance(dfs, pandas.DataFrame):
             dfs = [dfs]
+        if dfs:
+            for df in dfs:
+                self._validate_fields(df=df, key_field=key_field, id_field=id_field)
         self._dfs = dfs
         self._key_field = key_field
         self._id_field = id_field
 
+    def _get_key(self, body):
+        key = None
+        if self._key_field:
+            if isinstance(self._key_field, list):
+                key = []
+                for key_field in self._key_field:
+                    single_key = self._get_by_field_or_index(field=key_field, body=body)
+                    if pd.isna(single_key):
+                        return single_key, key_field
+                    key.append(single_key)
+
+            else:
+                key = self._get_by_field_or_index(field=self._key_field, body=body)
+                if pd.isna(key):
+                    return key, self._key_field
+        return key, None
+
+    def _get_element(self, body: OrderedDict):
+        return dict(body)
```

Comment:
As I wrote, I want to drop `OrderedDict`, but if we end up keeping it, why does it need to be converted to a regular dictionary here?

Reply:
we need to save the order of the dict, so _get_element in CSVSource will be able to access to the indexes of the columns as keys.